datafusion/sql/src/planner.rs

```diff
@@ -2453,6 +2453,17 @@ fn remove_join_expressions(
                     (_, Some(rr)) => Ok(Some(rr)),
                     _ => Ok(None),
                 }
+            },
+            // Fix for issue#78 join predicates from inside of OR expr also pulled up properly.
+            Operator::Or => {
```

Comment:
I don't think we can pull out `or`s the same way we would part of an `and` - altering a disjunction will affect the outcome. i.e. if it was `l.partkey=s.partkey or l.price < 5.00` and we extract the first part into a join clause, we will now only see ones with prices < 5, not ones that matched or were less than 5. A contrived example I'm sure, but I think I'd have to see a counter-example working before I am confident we didn't alter the results.

I think a better place to solve this might be with an optimizer rule, instead of in the planner - in an optimizer rule we could always just return `Err()` if the situation isn't perfectly to our liking.

Reply:
yes, I agree that OR predicates need to be handled differently. 
OR predicates are handled in below code.
          
            ` Operator::Or => {
                let mut left_join_keys = vec![];
                let mut right_join_keys = vec![];

                extract_possible_join_keys(left, &mut left_join_keys)?;
                extract_possible_join_keys(right, &mut right_join_keys)?;

                intersect( accum, &left_join_keys, &right_join_keys)
            }`            
In case of OR predicates, we pull the predicates if both left and right child are having same JOIN predicates. 
insect() function would ensure that we don't pull the all OR predicates.

In the example given above, `l.partkey=s.partkey or l.price < 5.00`, the left child of OR predicate would be `l.partkey=s.partkey` and right child of OR predicate would be ` l.price < 5.00'. There is no common predicates on left and right child of OR expr and so these predicates are not pulled to Join Predicates.

Reply:
We should not pull the predicates from disjunctions unless both left and right child of OR expr are having common conjunctions, since it will affect the query results. The idea for this fix is, "If there are common predicates in left and right child of OR expression then we should move those predicates to Join predicates. In case of Q19, it has common join predicate in all 3 OR expressions and so that Join predicate is pulled up to JOIN node.

Reply:
In general, I think this kind of rewrite should be done in an SQL optimizer pass so that it will apply to any query plan (e.g. that came from the DataFrame API) rather than only SQL